jupyterhubutils/lsstmgr/volumemanager.py

```diff
@@ -0,0 +1,179 @@
+'''Class to provide support for document-driven Volume assignment.
+'''
+
+import base64
+import json
+import os
+import yaml
+
+from ..utils import make_logger
+
+from kubernetes import client
+
+
+class LSSTVolumeManager(object):
+    volume_list = []
+    k8s_volumes = []
+    k8s_vol_mts = []
+
+    def __init__(self, *args, **kwargs):
+        self.log = make_logger()
+        self.log.debug("Creating LSSTVolumeManager")
+        self.parent = kwargs.pop('parent')
+
+    def make_volumes_from_config(self):
+        '''Create volume definition representation from document.
+        Override this in a subclass if you like.
+        '''
+        vollist = []
+        config = []
+        config_file = self.parent.config.volume_definition_file
+        if not os.path.exists(config_file):
+            return vollist
+        with open(config_file, "r") as fp:
+            config = json.load(fp)
+        for mtpt in config:
+            self.log.debug("mtpt: %r" % mtpt)
+            mountpoint = mtpt["mountpoint"]  # Fatal error if it doesn't exist
+            if mtpt.get("disabled"):
+                self.log.debug("Skipping disabled mountpoint %s" % mountpoint)
+                continue
+            if mountpoint[0] != "/":
+                mountpoint = "/" + mountpoint
+            host = (mtpt.get("fileserver-host") or
+                    self.parent.config.fileserver_host)
+            export = mtpt.get("fileserver-export") or (
+                "/exports" + mountpoint)
+            mode = (mtpt.get("mode") or "ro").lower()
+            k8s_vol = mtpt.get("kubernetes-volume")
+            if k8s_vol:
+                raise ValueError("Shadow PVs and matching PVCs " +
+                                 "are no longer supported!")
+            hostpath = mtpt.get("hostpath")
+            vollist.append({
+                "mountpoint": mountpoint,
+                "hostpath": hostpath,
+                "host": host,
+                "export": export,
+                "mode": mode,
+            })
+        self.log.debug("Volume list: %r" % vollist)
+        self.volume_list = vollist
+        volumes, mtpts = self._make_k8s_object_representations()
+        self.k8s_volumes = volumes
+        self.k8s_vol_mts = mtpts
+
+    def _make_k8s_object_representations(self):
+        volumes = []
+        mtpts = []
+        for vol in self.volume_list:
+            k8svol, k8smt = self._make_k8s_vol_objs(vol)
+            if k8svol:
+                volumes.append(k8svol)
+                mtpts.append(k8smt)
+        return (volumes, mtpts)
+
+    def _make_k8s_vol_objs(self, vol):
+        k8svol = None
+        k8smt = None
+        if vol.get("hostpath"):
+            k8svol = self._make_k8s_hostpath_vol(vol)
+            k8smt = self._make_k8s_hostpath_mt(vol)
+        else:
+            k8svol = self._make_k8s_nfs_vol(vol)
+            k8smt = self._make_k8s_nfs_mt(vol)
+        return k8svol, k8smt
```

Comment:
maybe collapse this into the function above and just append to k8s_volumes and k8s_vol_mts instead of returning it